### PR TITLE
Add interpreter for ConvolutionOp

### LINKED ISSUE #970: Add interpreter for convolution
Following the [testing guidelines](https://github.com/openxla/stablehlo/blob/main/docs/interpreter_checklist.md)

### PR DESCRIPTION
**Pull request by ghpvnist**
Since the interpreter implementation is already a mouthful, I've split #1314 to separate the implementation from the remaining checklist items. Once this is merged, I'll work on the remaining to update the rest of the docs.

closes #970

### CODE REVIEW

File: stablehlo/tests/interpret_convolution.mlir [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+    }
+  : (tensor<1x4x4x1xi64>, tensor<1x2x1x4xi64>) -> tensor<1x1x2x4xi64>
+  check.expect_eq_const %result, dense<[[[[1, 3, 10, 12],
+                                          [5, 7, 14, 16]]]]> : tensor<1x1x2x4xi64>
+  func.return
+}
+
+// -----
+
+func.func @convolution_feature_group_count_4() {
```
Comment (mlevesquedion) [inline stablehlo/tests/interpret_convolution.mlir:57]: The `feature_group_count` seems to be equal to 2. Typo?

File: stablehlo/tests/interpret_convolution.mlir [hunk 2]
```diff
@@ -0,0 +1,79 @@
+// RUN: stablehlo-translate --interpret -split-input-file %s
+
+func.func @convolution_op_test_si64() {
```
Comment (mlevesquedion) [inline stablehlo/tests/interpret_convolution.mlir:3]: Is the difference between `si64` and `i64` meaningful here? (I think `i64` in an MLIR type means `si64`. Is this relevant to point out in the test name?)

File: stablehlo/tests/interpret_convolution.mlir [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+  %rhs = stablehlo.constant dense<1> : tensor<3x3x1x1xi64>
+  %result = stablehlo.convolution(%lhs, %rhs)
+    dim_numbers = [b, 0, 1, f]x[0, 1, i, o]->[b, 0, 1, f],
+    window = {
+      stride = [4, 4],
+      lhs_dilate = [2, 2]
+    } {
+      batch_group_count = 1 : i64,
+      feature_group_count = 1 : i64,
+      precision_config = [#stablehlo<precision DEFAULT>, #stablehlo<precision DEFAULT>]
```
Comment (mlevesquedion) [inline stablehlo/tests/interpret_convolution.mlir:19]: Can this be elided? Since you're using the defaults. If not, we should consider using [DefaultValuedAttr](http://google3/third_party/llvm/llvm-project/mlir/include/mlir/IR/CommonAttrConstraints.td;l=93;rcl=602520726) (in a future change)

File: stablehlo/reference/Index.h [hunk 4]
```diff
@@ -121,6 +121,8 @@ using Index = Sizes;
 class IndexSpaceIterator {
  public:
   /// \name Constructor
+  IndexSpaceIterator(Sizes shape) : shape_(shape) { index_ = std::nullopt; }
```
Comment (mlevesquedion) [inline stablehlo/reference/Index.h:124]: Nit: Would it be possible to use a default value of `std::nullopt` for the `index` in the constructor below? I personally think that's cleaner but up to you.

File: stablehlo/reference/Ops.cpp [hunk 5]
```diff
@@ -154,6 +188,65 @@ SmallVector<SmallVector<uint32_t>> getReplicaGroups(
...
+          /*padding=*/
+          getDenseIntElementsAttr(
+              IntegerType::get(lhs.getType().getContext(), 64), paddingVector,
+              SmallVector<int64_t>(padding.size(), 2)),
+          lhsDilation, rhsDilation, windowReversal, inputBatchDimension,
+          inputFeatureDimension, ArrayRef<int64_t>(inputSpatialDimensions),
+          kernelInputFeatureDimension, kernelOutputFeatureDimension,
+          ArrayRef<int64_t>(kernelSpatialDimensions), outputBatchDimension,
+          outputFeatureDimension, ArrayRef<int64_t>(outputSpatialDimensions),
+          /*featureGroupCount=*/1, batchGroupCount,
```
Comment (mlevesquedion) [inline stablehlo/reference/Ops.cpp:220]: Why are we hard coding the feature group count here? Add a comment?

File: stablehlo/reference/Ops.cpp [hunk 6]
```diff
@@ -154,6 +188,65 @@ SmallVector<SmallVector<uint32_t>> getReplicaGroups(
...
+          /*precisionConfig=*/{}, inferredConvolutionType)))
```
Comment (mlevesquedion) [inline stablehlo/reference/Ops.cpp:221]: I'm assuming we don't need the precision config to infer the types?

File: stablehlo/reference/Ops.cpp [hunk 7]
```diff
@@ -372,6 +465,52 @@ SmallVector<InterpreterValue> eval(Region &region,
...
+          convolutionOp.getDimensionNumbers().getInputFeatureDimension(),
+          Axes(convolutionOp.getDimensionNumbers().getInputSpatialDimensions()),
+          convolutionOp.getDimensionNumbers().getKernelInputFeatureDimension(),
+          convolutionOp.getDimensionNumbers().getKernelOutputFeatureDimension(),
+          Axes(
+              convolutionOp.getDimensionNumbers().getKernelSpatialDimensions()),
+          convolutionOp.getDimensionNumbers().getOutputBatchDimension(),
+          convolutionOp.getDimensionNumbers().getOutputFeatureDimension(),
+          Axes(
+              convolutionOp.getDimensionNumbers().getOutputSpatialDimensions()),
```
Comment (mlevesquedion) [inline stablehlo/reference/Ops.cpp:510]: Nit: extract a "dimensionNumbers" variable?

File: stablehlo/reference/Ops.cpp [hunk 8]
```diff
@@ -372,6 +465,52 @@ SmallVector<InterpreterValue> eval(Region &region,
...
       auto result = evalConvertOp(operand, convertOp.getType());
       scope.add(convertOp.getResult(), result);
+    } else if (auto convolutionOp = dyn_cast<ConvolutionOp>(op)) {
+      auto lhs = scope.findTensor(convolutionOp.getLhs());
+      auto rhs = scope.findTensor(convolutionOp.getRhs());
+      auto rank = lhs.getRank();
+
+      SmallVector<int64_t> windowStrides(rank - 2, 1);
+      if (auto windowStridesAttr = convolutionOp.getWindowStridesAttr())
+        windowStrides = SmallVector<int64_t>(windowStridesAttr.asArrayRef());
```
Comment (mlevesquedion) [inline stablehlo/reference/Ops.cpp:475]: Nit: I think `auto windowStrides = convolutionOp.getWindowStrides().value_or(SmallVector<int64_t, 0>{})` should work.

(Actually it's unclear to me that `SmallVector` is needed here, you could do something similar with `ArrayRef`.)

**Review by mlevesquedion — COMMENTED** [review]
Took a quick look, will let someone else do a full review.

File: stablehlo/reference/Ops.cpp (hunk 6 above)
Comment (ghpvnist) [reply stablehlo/reference/Ops.cpp:221]: Yes that's right.

File: stablehlo/reference/Ops.cpp (hunk 7 above)
Comment (ghpvnist) [reply stablehlo/reference/Ops.cpp]: Done.

File: stablehlo/reference/Ops.cpp (hunk 8 above)
Comment (ghpvnist) [reply stablehlo/reference/Ops.cpp:475]: We want default value to be filled with ones. `value_or` actually returns an `ArrayRef`, so there would be a type mismatch (and creating defaults with `ArrayRef` doesn't offer more readability compared to using `SmallVector` for the type).

File: stablehlo/reference/Index.h (hunk 4 above)
Comment (ghpvnist) [reply stablehlo/reference/Index.h:124]: Done.

File: stablehlo/tests/interpret_convolution.mlir (hunk 3 above)
Comment (ghpvnist) [reply stablehlo/tests/interpret_convolution.mlir]: The interpreter currently doesn't handle different variations of `precision_config`, so this makes it explicit for now. There's also a couple of issues (#445, #755, #879) to figure out the exact story for `precision_config`.

File: stablehlo/tests/interpret_convolution.mlir (hunk 2 above)
Comment (ghpvnist) [reply stablehlo/tests/interpret_convolution.mlir]: There's a long standing issue #22 to try to converge to one. Long story short, the types were inherited from MHLO when StableHLO was first created. To avoid ambiguity, we explicitly write `si64` in the tests.

File: stablehlo/tests/interpret_convolution.mlir (hunk 1 above)
Comment (ghpvnist) [reply stablehlo/tests/interpret_convolution.mlir]: Done.

**Review by GleasonK — APPROVED** [review]

File: stablehlo/reference/Index.h [hunk 9]
```diff
@@ -121,9 +121,11 @@ using Index = Sizes;
...
+
   IndexSpaceIterator(Sizes shape, std::optional<Index> index)
-      : shape_(shape), index_(index) {
-    if (index && !index->inBounds(shape)) index_ = std::nullopt;
+      : shape_(shape), index_(std::nullopt) {
+    if (index && index->inBounds(shape)) index_ = index;
```
[the remark this answers is not in the document]
Comment (ghpvnist) [reply stablehlo/reference/Index.h:128]: One use case is for comparing against end of the iterator. It's useful when iterating through a tensor one by one.

File: stablehlo/reference/Ops.cpp [hunk 10]
```diff
@@ -54,6 +54,27 @@ Index evalIndex(Tensor tensor) {
...
 
+Tensor evalDotGeneralOp(const Tensor &lhs, const Tensor &rhs,
+                        const Axes &lhsBatchingDimensions,
+                        const Axes &rhsBatchingDimensions,
+                        const Axes &lhsContractingDimensions,
+                        const Axes &rhsContractingDimensions) {
+  SmallVector<ShapedTypeComponents> inferredDotGeneralType;
+  if (failed(hlo::inferDotGeneralOp(
+          /*location=*/{}, lhs.getType(), rhs.getType(),
+          /*lhsBatchingDimensions=*/{}, /*rhsBatchingDimensions*/ {},
```
[the remark this answers is not in the document]
Comment (ghpvnist) [reply stablehlo/reference/Ops.cpp:63]: The only use of this function is by convolution and convolution doesn't use batching dimension. Since they aren't used, I've removed it.
